src/wp-includes/blocks.php

```diff
@@ -1126,15 +1126,39 @@ function build_query_vars_from_query_block( $block, $page ) {
 			$query['offset']         = ( $per_page * ( $page - 1 ) ) + $offset;
 			$query['posts_per_page'] = $per_page;
 		}
-		if ( ! empty( $block->context['query']['categoryIds'] ) ) {
-			$term_ids              = array_map( 'intval', $block->context['query']['categoryIds'] );
-			$term_ids              = array_filter( $term_ids );
-			$query['category__in'] = $term_ids;
+		// Migrate `categoryIds` and `tagIds` to `tax_query` for backwards compatibility.
+		if ( ! empty( $block->context['query']['categoryIds'] ) || ! empty( $block->context['query']['tagIds'] ) ) {
+			$tax_query = array();
+			if ( ! empty( $block->context['query']['categoryIds'] ) ) {
+				$tax_query[] = array(
+					'taxonomy'         => 'category',
+					'terms'            => $block->context['query']['categoryIds'],
+					'include_children' => false,
+				);
+			}
+			if ( ! empty( $block->context['query']['tagIds'] ) ) {
+				$tax_query[] = array(
+					'taxonomy'         => 'post_tag',
+					'terms'            => $block->context['query']['tagIds'],
+					'include_children' => false,
+				);
+			}
+			$query['tax_query'] = $tax_query;
 		}
-		if ( ! empty( $block->context['query']['tagIds'] ) ) {
-			$term_ids         = array_map( 'intval', $block->context['query']['tagIds'] );
-			$term_ids         = array_filter( $term_ids );
-			$query['tag__in'] = $term_ids;
+		if ( ! empty( $block->context['query']['taxQuery'] ) ) {
+			$query['tax_query'] = array();
+			foreach ( $block->context['query']['taxQuery'] as $taxonomy => $terms ) {
+				if ( ! empty( $terms ) ) {
+					$term_ids = array_map( 'intval', $terms );
+					$term_ids = array_filter( $term_ids );
+
+					$query['tax_query'][] = array(
+						'taxonomy'         => $taxonomy,
+						'terms'            => $terms,
```

Comment:
`$taxonomy` and `$terms` are unescaped here. The taxonomy also needs to be confirmed as publicly queryable.


```php
if ( ! is_taxonomy_viewable( $taxonomy ) { continue } // This runs it through `taxonomy_exists()` too
```

* If term is a slug, then `array_filter( array_map( 'sanitize_title', $terms ) )`
* If this is a typo and intended to be term_ids, then do `array_filter( array_map( 'intval', $terms ) )` here so the value isn't accidently messed up by our future selves by adding an error on line 1154.

Reply:
It's only `ids`, there is no support for `slugs`. I'll see about the validations. I'll make some updates to [the GB PR](https://github.com/WordPress/gutenberg/pull/39970) and will ping you as well then to take a look.

Reply:
Also noting that the `$terms` usage was a bug and will be replaced with `$term_ids`, which are sanitized.

Reply:
Resolved with https://github.com/WordPress/gutenberg/pull/39970 in https://github.com/WordPress/wordpress-develop/commit/e51fd0c51ea297bbc717cb42b344b956d0f3f8b4.